### PR TITLE
router: fix request retries with bodies

### PR DESCRIPTION
**Pull request by bgentry**
As reported in #872, http.Transport will call Close() on any Request.Body after it encounters any error. We are retrying the same request under certain circumstances (dial errors in particular), so we need to keep the body open until we're done retrying.

Adds some test coverage for the error case reported in #872 by also testing `POST` requests with bodies, and fixes the underlying issue as well.

### CODE REVIEW

File: router/http.go [hunk 1]
```diff
@@ -491,9 +492,11 @@ func (s *httpService) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 				continue
 			}
 			log.Println("http: proxy error:", err)
+			outreq.Body.(*fakeCloseReadCloser).RealClose()
 			fail(w, 503)
 			return
 		}
+		outreq.Body.(*fakeCloseReadCloser).RealClose()
```
Comment (bgentry) [inline router/http.go]: I'm not super happy with the way this ends up being duplicated here, but didn't think it was worth the effort to try and factor out this retry loop in light of our other pending PR (#842). Maybe we should just make sure it's done more cleanly there?

Comment (titanous) [inline router/http.go]: Yeah, fixing this in the refactor is fine. Thanks for the quick bugfix!

**Comment by titanous** [comment]
LGTM.

**Comment by bgentry** [comment]
oh /cc @benburkert

**Comment by bamarni** [comment]
I'm coming from a linked issue which is similar in traefik (https://github.com/containous/traefik/issues/1008) but I'm not sure this patch is enough.

It seems like even though subsequent reads wouldn't trigger errors, they wouldn't read anything either, or do I miss something?

**Comment by titanous** [comment]
@bamarni If I remember correctly, the problem this was working around was the body being closed by `net/http` during a failed dial, before anything was read, so turning the close into a no-op is a valid solution.

**Comment by bamarni** [comment]
@titanous : makes sense, thanks for the feedback.

**Pull request merged**